canonical serialization returns 0.5 instead of calc(0.5)

colelao opened the pull request:
#### 8dd3b24cc3bd8ccc3422cb201aa114d41cb5162c
<pre>
canonical serialization returns 0.5 instead of calc(0.5)
<a href="https://bugs.webkit.org/show_bug.cgi?id=294363">https://bugs.webkit.org/show_bug.cgi?id=294363</a>
<a href="https://rdar.apple.com/153784390">rdar://153784390</a>

Reviewed by NOBODY (OOPS!).

There was a work around dealing with the opacity serialization quirk. The work around incorrectly serialized the value, so this fixes it.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::serializeLonghandValue):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8dd3b24cc3bd8ccc3422cb201aa114d41cb5162c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113628 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33335 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23775 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119799 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33942 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41905 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86421 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41522 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27089 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102107 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66756 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26332 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20228 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63529 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96466 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20318 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123034 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40631 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30341 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95269 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41022 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98323 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95016 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40149 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17947 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36867 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40514 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46033 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40175 "Built successfully") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41930 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nullhook commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -71,8 +71,13 @@ String serializeLonghandValue(const CSS::SerializationContext& context, CSSPrope
+            result.append("calc("_s);
+            result.append(makeString(primitive->resolveAsPercentageDeprecated() / 100));
+            result.append(")"_s);
+            return result.toString();
```
consider, if possible?

```c++ 
makeString("calc(", primitive->resolveAsPercentageDeprecated() / 100, ")")
```

may want to include `<wtf/text/StringConcatenateNumbers.h>` ?

colelao committed:
canonical serialization returns 0.5 instead of calc(0.5)
https://bugs.webkit.org/show_bug.cgi?id=294363
rdar://153784390

Reviewed by NOBODY (OOPS!).

There was a work around dealing with the opacity serialization quirk. The work around incorrectly serialized the value, so this fixes it.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::serializeLonghandValue):

nullhook commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -71,8 +71,11 @@ String serializeLonghandValue(const CSS::SerializationContext& context, CSSPrope
-        if (auto* primitive = dynamicDowncast<CSSPrimitiveValue>(value); primitive && primitive->isPercentage())
+        if (auto* primitive = dynamicDowncast<CSSPrimitiveValue>(value); primitive && primitive->isPercentage()) {
+            if (primitive->isCalculated())
+                return makeString("calc("_s, primitive->resolveAsPercentageDeprecated() / 100, ")"_s);
```
this assumes that all calculated values are `calc()` functions.

what about other math functions? for ex. `opacity: clamp(..);` will this return "calc(" ?

nullhook replied:
...i'm curious how you're passing test for other function types

nullhook reviewed (approved):
lgtm

nullhook reviewed (changes requested):
we may need to think about this more...

as per the spec:
> If the [calculation tree’s](https://drafts.csswg.org/css-values/#calculation-tree) root node is a numeric value, or a [calc-operator node](https://drafts.csswg.org/css-values/#calculation-tree-calc-operator-nodes), let s be a string initially containing "calc(". Otherwise, let s be a string initially containing the name of the root node, lowercased (such as "sin" or "max"), followed by a "(" (open parenthesis).

this is how i'm interpreting this as... 

`50% + 10%` should be serialized as `calc(`
`clamp(50%, 0%, 70%)` should be serialized as `clamp(`

looks like `/css/calc/CSSCalcTree+Serialization.cpp` has the machinery but i'm not sure if it's ready to be used yet? 

cc @weinig
